packages/live-share-inking/src/canvas/DryWetCanvas.ts

```diff
@@ -0,0 +1,269 @@
+
+import { InkingCanvas } from "./InkingCanvas";
+import { getPressureAdjustedSize, IPointerPoint, DefaultPenBrush, IBrush, toCssColor } from "../core";
+import { computeQuadBetweenTwoCircles, computeQuadBetweenTwoRectangles, IQuadPathItem } from "../core/Internals";
+
+/**
+ * Represents the base class from wet and dry canvases, implementing the common rendering logic.
+ */
+export abstract class DryWetCanvas extends InkingCanvas {
+    private _innerLayer?: CanvasRenderingContext2D;
+    private _pendingPointsStartIndex = 0;
+    private _points: IPointerPoint[] = [];
+
+    private computeQuadPath(tipSize: number): IQuadPathItem[] {
+        const result: IQuadPathItem[] = [];
+        const tipHalfSize = tipSize / 2;
+
+        if (this._pendingPointsStartIndex < this._points.length) {
+            let previousPoint: IPointerPoint | undefined = undefined;
+            let previousPointPressureAdjustedTip = 0;
+
+            if (this._pendingPointsStartIndex > 0) {
+                previousPoint = this._points[this._pendingPointsStartIndex - 1];
+                previousPointPressureAdjustedTip = getPressureAdjustedSize(tipHalfSize, previousPoint.pressure);
+            }
+
+            for (let i = this._pendingPointsStartIndex; i < this._points.length; i++) {
+                const p = this._points[i];
+
+                let pressureAdjustedTip = getPressureAdjustedSize(tipHalfSize, p.pressure);
+
+                const pathItem: IQuadPathItem = {
+                    endPoint: p,
+                    tipSize: pressureAdjustedTip
+                };
+
+                if (previousPoint !== undefined) {
+                    pathItem.quad = this.brush.tip === "ellipse"
+                        ? computeQuadBetweenTwoCircles(
+                            p,
+                            pressureAdjustedTip,
+                            previousPoint,
+                            previousPointPressureAdjustedTip)
+                        : computeQuadBetweenTwoRectangles(
+                            p,
+                            pressureAdjustedTip,
+                            pressureAdjustedTip,
+                            previousPoint,
+                            previousPointPressureAdjustedTip,
+                            previousPointPressureAdjustedTip);
+                }
+
+                result.push(pathItem);
+
+                previousPoint = p;
+                previousPointPressureAdjustedTip = pressureAdjustedTip;
+            }
+        }
+
+        return result;
+    }
+
+    private renderQuadPath(context: CanvasRenderingContext2D, path: IQuadPathItem[]) {
+        const cssColor = this.getBrushCssColor();
+
+        context.strokeStyle = cssColor;
+        context.fillStyle = cssColor;
+
+        context.beginPath();
+
+        for (let item of path) {
+            if (item.quad !== undefined) {
+                this.renderQuad(context, item.quad);
+            }
+
+            if (this.brush.tip === "ellipse") {
```

Comment:
Maybe BrushTipShape?

Reply:
(and other places, if you are open to the change)

Reply:
Same as above. Brush.tip is of type BrushTipShape, it's just that BrushTipShape is a type, not an enum.